File: DFS/047.Permutations-II/047.Permutations-II.ts
function permuteUnique(nums: number[]): number[][] {
  nums.sort((a, b) => a - b);
  const ret: number[][] = [];
  let used = new Array<boolean>(nums.length).fill(false);
  let current: number[] = [];
  dfs();
  return ret;
  function dfs() {
    if (current.length === nums.length) {
      ret.push([...current]);
      return;
    }

    for (let i = 0; i < nums.length; i++) {
      if (used[i]) {
        continue;
      }

      if (i < nums.length && nums[i + 1] === nums[i] && !used[i + 1]) {
        continue;
      }

      used[i] = true;
      current.push(nums[i]);
      dfs();
      current.pop();
      used[i] = false;
    }
  }
}

File: DFS/047.Permutations-II/047.Permutations-II.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        result = []
        current = []
        used = [False] * len(nums)

        def dfs():
            if len(current) == len(nums):
                result.append(current[:])
                return
            
            chosen = set()

            for i in range(len(nums)):
                if nums[i] not in chosen and not used[i]:
                    chosen.add(nums[i])

                    used[i] = True
                    current.append(nums[i])
                    dfs()
                    current.pop()
                    used[i] = False
        
        dfs()
        return result
            
